### PR TITLE
Fix #133: Add scalafmt

### LINKED ISSUE #133: Use scalafmt

### PR DESCRIPTION
**Pull request by jiminhsieh**
Fix #133 
When we execute `compile` and `test`, it will format `main` and `test` separately. To format `*.sbt` and `project/*.scala`, we have to execute `scalafmtSbt`.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -9,15 +9,18 @@ lazy val commonSettings = Seq(
...
 
 /** We use https://github.com/DavidGregory084/sbt-tpolecat but some of these are broken */
-def ignoreScalacOptions(scalaVersion: String): Seq[String] = CrossVersion.partialVersion(scalaVersion) match {
-  case Some((2, 10)) => Seq("-Ywarn-numeric-widen") // buggy in 2.10
-  case Some((2, 11)) => Seq("-Ywarn-value-discard") // This is broken in 2.11 for Unit types
-  case _             => Nil
-}
+def ignoreScalacOptions(scalaVersion: String): Seq[String] =
+  CrossVersion.partialVersion(scalaVersion) match {
+    case Some((2, 10)) => Seq("-Ywarn-numeric-widen") // buggy in 2.10
```
Comment (pathikrit) [inline build.sbt]: what happened here? Line 18 and 19 looks different?

File: benchmarks/src/test/scala/better/files/ScannerBenchmark.scala [hunk 2]
```diff
@@ -47,7 +48,8 @@ class ScannerBenchmark extends Benchmark {
           |Hello World
           |19
         """.stripMargin
-      val scanner = scannerBuilder(new BufferedReader(new StringReader(canaryData)))
+      val scanner =
```
Comment (pathikrit) [inline benchmarks/src/test/scala/better/files/ScannerBenchmark.scala]: what's wrong here? line length? can we increase it to 120/160 chars?

File: project/plugins.sbt [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
-addSbtPlugin("org.scoverage"              %   "sbt-scoverage"           % "1.5.1")
-addSbtPlugin("com.timushev.sbt"           %   "sbt-updates"             % "0.3.1")
-addSbtPlugin("com.typesafe.sbt"           %   "sbt-ghpages"             % "0.6.2")
-addSbtPlugin("com.updateimpact"           %   "updateimpact-sbt-plugin" % "2.1.3")
-addSbtPlugin("org.xerial.sbt"             %   "sbt-sonatype"            % "2.0")
-addSbtPlugin("com.codacy"                 %   "sbt-codacy-coverage"     % "1.3.11")
-addSbtPlugin("io.github.davidgregory084"  %   "sbt-tpolecat"            % "0.1.3")
+addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.4.1")
+addSbtPlugin("com.github.gseitz" % "sbt-release" % "1.0.6")
+addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.1.0")
```
Comment (pathikrit) [inline project/plugins.sbt]: Not a big deal but I liked the previous alignment :)

**Comment by jiminhsieh** [comment]
Thanks for your review!
> 1. Can you add the scalafmt config file so we can tweak some settings?

Sure. 
> 2. Looks like in certain cases, the formatting earlier was much easier to read than current. I think it is perhaps the line length? Can we increase it? After you increase it revert all code to master and rerun the formatter?

I will try to increase to 120 or 160 chars than we can see is it better than default column.
> 3. Can you split out the commit (use git rebase -i) into two commits - a) that introduces scalafmt b) the commit that reformats. This way we can revert b) and play with the settings.

Yes, I agree. That's why I did two commits.
> 4. Can you add a sbt task called sbt formatAll which reformats all code (including test, project/*.scala)?
> 5. Can you make sbt compile depend on sbt formatAll i.e. compile always reformats code?
> 6. Can you add a task sbt checkFormat that checks if code is formatted?
> 7. Can you make it so that sbt test depends on sbt checkFormat i.e. test fails immediately if code is not formatted. Normally this would be impossible since sbt test depends on sbt compile which reformats code.

I will try to make those happen. But I have to make sure this formatting works as your expectation. In the meantime, I can do those features locally. So I will reset the previous commit and add `.scalafmt.conf` with 120 or 160 chars.

File: project/plugins.sbt (hunk 3 above)
Comment (jiminhsieh) [reply project/plugins.sbt]: Yeah, I did like the previous alignment! I will try to find a way to keep original alignment.

**Comment by jiminhsieh** [comment]
@pathikrit I finished the requirements. [This](https://circleci.com/gh/jiminhsieh/better-files/137) is the failed result of before I formated the source code.

**Comment by pathikrit** [comment]
@jiminhsieh : This looks awesome. Some minor comments:

1) Can we make the implicit params be in new lines somehow if it is too long e.g.:
```scala
// Original
def foo(param: X)(implicit x: Eq): Foo = //Long line of signature

// Current:
def foo(param: X)(
 implicit x: Eq): Foo =

// Intended:
def foo(param: X)
             (implicit x: Eq): Foo =
```

2. Is there any config to do import re-ordering? I would like 4 groups:
Group 1: better.file imports
Group 2: any external deps
Group 3: Java imports
Group 4: Scala imports
Each group sorted and separated from other group by blank line

But, this looks great! Thanks for this!

**Comment by jiminhsieh** [comment]
About implicit parms, scalafmt only supports two types of format: 
```scala
// newlines.afterImplicitKWInVerticalMultiline = true
// newlines.beforeImplicitKWInVerticalMultiline = true
def format(
  code: String,
  age: Int
)(
  implicit
  ev: Parser,
  c: Context
): String
``` 
and
```scala
// newlines.afterImplicitKWInVerticalMultiline = true
// newlines.beforeImplicitKWInVerticalMultiline = false
def format(
  code: String,
  age: Int
)(implicit
  ev: Parser,
  c: Context
): String
```

About import re-ordering, I saw this [comment](https://github.com/scalameta/scalafmt/issues/722#issuecomment-354665431). We have to introduce **scalafix**. I can take a look at scalafix latter and see does it support this feature.

**Comment by pathikrit** [comment]
@jiminhsieh : Let's chose this:
```
// newlines.afterImplicitKWInVerticalMultiline = true
// newlines.beforeImplicitKWInVerticalMultiline = false
```

And we can do scalafix for imports in separate PR. Thanks and I will merge after this.

**Comment by jiminhsieh** [comment]
@pathikrit I patched the new config. If the last 2 commits are good to you, I will sqush all of commits to 2 commits.

**Comment by pathikrit** [comment]
This looks great! :shipit:

**Comment by jiminhsieh** [comment]
Done. :)

**Pull request merged**